File: Python/ByRound/1977/1977_D_XORificator.py
# Problem: CF 1977 D - XORificator
# https://codeforces.com/contest/1977/problem/D

"""
Algorithm: XORificator
Purpose: Given a binary n×m matrix, determine the maximum number of columns that can be made "special" (exactly one 1) by choosing which rows to XOR (invert all values in that row). 
Technique: 
- Use randomization with XOR to reduce problem to counting similar XOR patterns
- For each column, compute XOR of row indices where that column has a 1
- Use this as a hash to count how many columns match a certain XOR pattern (same pattern implies same result after row flips)
- Find the best row/column pair to use for flipping

Time Complexity: O(n * m), since we process each element once in the main loop
Space Complexity: O(n + m), for storing the matrix and auxiliary data structures
"""

import os
import sys
from io import BytesIO, IOBase
from random import getrandbits

from _collections import defaultdict

BUFSIZE = 8192


class FastIO(IOBase):
    newlines = 0

    def __init__(self, file):
        self._fd = file.fileno()
        self.buffer = BytesIO()
        self.writable = "x" in file.mode or "r" not in file.mode
        self.write = self.buffer.write if self.writable else None

    def read(self):
        while True:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            if not b:
                break
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines = 0
        return self.buffer.read()

    def readline(self):
        while self.newlines == 0:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            self.newlines = b.count(b"\n") + (not b)
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines -= 1
        return self.buffer.readline()

    def flush(self):
        if self.writable:
            os.write(self._fd, self.buffer.getvalue())
            self.buffer.truncate(0), self.buffer.seek(0)


class IOWrapper(IOBase):
    def __init__(self, file):
        self.buffer = FastIO(file)
        self.flush = self.buffer.flush
        self.writable = self.buffer.writable
        self.write = lambda s: self.buffer.write(s.encode("ascii"))
        self.read = lambda: self.buffer.read().decode("ascii")
        self.readline = lambda: self.buffer.readline().decode("ascii")


if sys.version_info[0] < 3:
    sys.stdin = BytesIO(os.read(0, os.fstat(0).st_size))
else:
    sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)

file = sys.stdin
if os.environ.get("USER") == "loic":
    file = open("data.in")

line = lambda: file.readline().split()
ui = lambda: int(line()[0])
mi = lambda: map(int, line())
li = lambda: list(mi())
lsi = lambda: list(map(int, line()[0]))


def solve():

    rd = []
    for i in range(N):   # Generate random bits for each row to use in hashing
        rd.append(getrandbits(63))

    mp = defaultdict(int)  # Map to count occurrences of XOR patterns

    mx = 0  # Maximum special columns achievable
    ref = (-1, -1)  # Best row and column indices to flip to achieve max
    for j in range(M):  # For each column
        x = 0  # XOR of row indices with 1s in current column
        for i in range(N):
            if A[i][j] == 1:
                x ^= rd[i]

        # Try flipping each row and count how many columns would be special
        for i in range(N):
            x ^= rd[i]  # Flip the bit at index i (simulate XOR operation)
            mp[x] += 1  # Count how many columns match this pattern
            if mp[x] > mx:  # Update best match if current is better
                mx = mp[x]
                ref = (i, j)
            x ^= rd[i]  # Undo the change

    res = []
    for i in range(N):
        # Determine if we need to flip this row based on best solution
        res.append(A[i][ref[1]] ^ (i == ref[0]))

    return str(mx) + "\n" + "".join(str(v) for v in res)


for test in range(1, ui() + 1):
    N, M = mi()
    A = [lsi() for _ in range(N)]

    print(solve())

file.close()


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/D/1977_D_XORificator.cpp
// Problem: CF 1977 D - XORificator
// https://codeforces.com/contest/1977/problem/D

/*
 * Problem: D. XORificator
 *
 * Algorithm/Technique: Hashing with XOR, Greedy approach
 *
 * Time Complexity: O(n * m * log(n * m)) due to sorting
 * Space Complexity: O(n * m) for storing hash values and pairs
 *
 * Approach:
 * 1. For each column, compute a hash value that represents the "signature" of that column.
 *    This signature is computed by XORing the hash values of all rows where the column has '1'.
 * 2. For each cell (i, j), we compute the new signature if we flip row i (inverting all values in that row).
 *    This gives us the potential new signature for column j after flipping row i.
 * 3. Sort all such (signature, column_index) pairs and find the maximum number of columns
 *    with the same signature, which indicates how many columns would be made special.
 * 4. Choose the best row flip configuration and output the result.
 */

#include <algorithm>
#include <cstdint>
#include <ctime>
#include <iostream>
#include <random>
#include <string>
#include <vector>
#include <utility>

template <typename T>
std::istream &operator>>(std::istream &input, std::vector<T> &v) {
  for (T &a : v)
    input >> a;
  return input;
}
void answer(unsigned x, const std::string &y) {
  std::cout << x << '\n';
  std::cout << y << '\n';
}
uint64_t h[300000]; // Precomputed hash values for rows
void initialize() {
  std::mt19937_64 g(time(nullptr));
  for (size_t i = 0; i < 300000; ++i)
    h[i] = g(); // Initialize random hashes
}
void solve(std::vector<std::string> &a) {
  const size_t n = a.size(), m = a[0].size();
  std::vector<std::pair<uint64_t, size_t>> s; // Store <signature, column_index> pairs
  
  // Generate the signature for each column without flipping any rows
  for (size_t j = 0; j < m; ++j) {
    uint64_t v = 0;
    for (size_t i = 0; i < n; ++i) {
      if (a[i][j] == '1')
        v ^= h[i]; // XOR the hash of row i if the cell is '1'
    }
    // For each cell (i, j), calculate what the new signature would be if we flip row i
    for (size_t i = 0; i < n; ++i)
      s.emplace_back(v ^ h[i], i * m + j); // Store the signature and column index
  }
  
  std::sort(s.begin(), s.end()); // Sort by signature to group same signatures together
  
  std::pair<unsigned, uint64_t> x = {0, 0}; // <max_count, index_of_best_column>
  
  // Iterate through sorted signatures to find maximum count of same signatures
  for (size_t i = 0, j = 0; i < s.size(); i = j) {
    while (j < s.size() && s[j].first == s[i].first)
      ++j; // Count all elements with the same signature
      
    if (j - i > x.first) {
      x.first = j - i; // Update max count
      x.second = s[i].second; // Store index of column that caused this count
    }
  }
  
  // Flip the row that makes the maximum number of columns special
  a[x.second / m][x.second % m] ^= 1; // Flip the row represented by x.second / m
  
  // Construct the result string indicating which rows were flipped
  std::string w;
  for (size_t i = 0; i < n; ++i)
    w.push_back(a[i][x.second % m]); // Since all columns are flipped, we need to check if we flipped row i
  
  answer(x.first, w);
}
void test_case() {
  size_t n, m;
  std::cin >> n >> m;
  std::vector<std::string> a(n);
  std::cin >> a;
  solve(a);
}
int main() {
  std::cin.tie(nullptr)->sync_with_stdio(false);
  initialize();
  size_t t;
  std::cin >> t;
  while (t-- > 0)
    test_case();
  return 0;
}


// https://github.com/VaHiX/CodeForces/